storm/src/main/storm/mesos/resources/RangeResource.java

```diff
@@ -0,0 +1,214 @@
+   */
+  @Override
+  public List<ResourceEntry> removeAndGet(RangeResourceEntry rangeResourceEntry, Comparator<ReservationType> reservationTypeCompartor) throws ResourceNotAvailabeException {
+    if (isAvailable(rangeResourceEntry)) {
+      List<ReservationType> reservationTypeList = Arrays.asList(ReservationType.values());
+      Collections.sort(reservationTypeList, reservationTypeCompartor);
+      return removeAndGet(reservationTypeList, rangeResourceEntry);
+    }
+
+    String message = String.format("ResourceType '%s' is not available. Requested value: %s Available: %s",
+                                   resourceType, rangeResourceEntry, toString());
+    throw new ResourceNotAvailabeException(message);
+  }
+
+  public String toString(List<RangeResourceEntry> ranges) {
+    List<String> rangeStrings = new ArrayList<>();
+    for (RangeResourceEntry range : ranges) {
+      String beginStr = String.valueOf(range.getBegin());
+      String endStr = String.valueOf(range.getEnd());
+      /*
+      * A Range representing a single number still has both Range.begin
+      * and Range.end populated, but they are set to the same value.
+      * In that case we just return "N" instead of "N-N".
+      */
+      if (range.getBegin() == range.getEnd()) {
+        rangeStrings.add(beginStr);
+      } else {
+        rangeStrings.add(String.format("%s-%s", beginStr, endStr));
+      }
+    }
+    return String.format("[%s]", StringUtils.join(rangeStrings, ","));
+  }
+
+  public String toString(ReservationType reservationType) {
+    return String.format("%s : %s", reservationType, toString(availableResourcesByReservationType.get(reservationType)));
+  }
+
+  public String toString() {
+    List<RangeResourceEntry> resourceRanges = new ArrayList<>();
+    for (Map.Entry<ReservationType, List<RangeResourceEntry>> entry : availableResourcesByReservationType.entrySet()) {
+      resourceRanges.addAll(entry.getValue());
+    }
+    return toString(resourceRanges);
+  }
+
+  private List<ResourceEntry> removeAndGet(Collection<ReservationType> reservationTypes, RangeResourceEntry rangeResourceEntry) {
+    List<ResourceEntry> removedResources = new ArrayList<>();
+    Long beginValue = rangeResourceEntry.getBegin();
+    Long endValue = rangeResourceEntry.getEnd();
+
+    for (ReservationType reservationType : reservationTypes) {
+      List<RangeResourceEntry> rangeResourceEntryList = availableResourcesByReservationType.get(reservationType);
+      for (int i = 0; i < rangeResourceEntryList.size(); i++) {
+        RangeResourceEntry tmp = rangeResourceEntryList.get(i);
+        if (beginValue >= tmp.getBegin() && endValue <= tmp.getEnd()) {
+          rangeResourceEntryList.remove(i);
+          // We already removed the entry. So when beginValue == endValue,
+          // we dont have to add a new entry
+          if (tmp.getBegin().equals(tmp.getEnd()) || (tmp.getBegin().equals(beginValue) && tmp.getEnd().equals(endValue))) {
```

Comment:
Why `.equals` and not `==`? (If there's no particular reason, I feel like `==` is slightly more readable.)